stream/harfanglab-intel/src/harfanglab_intel_connector/api_client.py

```diff
@@ -0,0 +1,379 @@
+            sigma_source_list = self._post_source("SigmaSource")
+
+        yara_source_list = self._get_source("YaraSource")
+        if not yara_source_list:
+            yara_source_list = self._post_source("YaraSource")
+
+        self.ioc_list_id = ioc_source_list["id"]
+        self.sigma_list_id = sigma_source_list["id"]
+        self.yara_list_id = yara_source_list["id"]
+
+    def _send_request(self, method: str, url: str, **kwargs) -> dict | None:
+        """
+        Send a request to Harfanglab API.
+        :param method: Request HTTP method
+        :param url: Request URL
+        :param kwargs: Any arguments valid for session.requests() method
+        :return: Any data returned by the API
+        """
+        try:
+            response = self.session.request(method, url, **kwargs)
+            response.raise_for_status()
+
+            self.helper.connector_logger.info(
+                f"[API] HTTP {method.upper()} Request to endpoint",
+                {"url_path": url},
+            )
+
+            if response.content:
+                return response.json()
+        except requests.RequestException as err:
+            self.helper.connector_logger.error(
+                "[API] Error while requesting : ",
+                {"url_path": f"{method.upper()} {url}", "error": str(err)},
+            )
+            return None
+
+    def _post_source(self, source_type: str) -> dict:
+        """
+        Create a source list of specified type on Harfanglab.
+        Source list name equals to HARFANGLAB_INTEL_SOURCE_LIST_NAME environment variable.
+        :param source_type: Type of source to create (either `IOCSource`, `SigmaSource` or `YaraSource`)
+        :return: Created source data
+        """
+        url = f"{self.api_base_url}/api/data/threat_intelligence/{source_type}"
+        body = {
+            "name": self.config.harfanglab_source_list_name,
+            "description": "Cyber Threat Intelligence knowledge imported from OpenCTI, and any changes must be made only to it.",
+            "enabled": True,
+        }
+
+        data = self._send_request(
+            method="post",
+            url=url,
+            json=body,
+        )
+        return data
+
+    def _get_source(self, source_type: str) -> dict:
+        """
+        Get a source list of specified type from Harfanglab.
+        Source list is returned according to HARFANGLAB_INTEL_SOURCE_LIST_NAME environment variable.
+        :param source_type: Type of source to create (either `IOCSource`, `SigmaSource` or `YaraSource`)
+        :return: Found source data
+        """
+        url = f"{self.api_base_url}/api/data/threat_intelligence/{source_type}"
+        params = {"name__exact": self.config.harfanglab_source_list_name}
+
+        data = self._send_request(
+            method="get",
+            url=url,
+            params=params,
+        )
+        results = data["results"]
+        if len(results):
+            return results[0]
```

Comment:
I take only the first element of the results list because the filter applied is "name__exact" so it means that Harfanglab API will either send back one result or none (I checked and lists' names can't be duplicated). That's why I didn't mind about raising an error, do you think it's useful in this particular case (only zero or one possible result)?

Typing updated 👍